flash/data/data_source.py

```diff
@@ -0,0 +1,263 @@
+        labels: Optional[Sequence[str]] = None
+    ):
+        super().__init__(
+            train_data=(train_inputs, train_targets),
+            val_data=(val_inputs, val_targets),
+            test_data=(test_inputs, test_targets),
+            predict_data=(predict_inputs, None),
+        )
+
+        self.labels = labels
+
+        if self.labels is not None:
+            self.preprocess.set_state(LabelsState(self.labels))
+
+    def load_data(self, data: Any, dataset: Optional[Any] = None) -> Iterable[Mapping[str, Any]]:
+        inputs, targets = data
+        if targets is None:
+            return [{'input': input} for input in inputs]
+        return [{'input': input, 'target': target} for input, target in zip(inputs, targets)]
+
+
+class FoldersDataSource(DataSource, ABC):
+
+    def __init__(
+        self,
+        train_folder: Optional[Union[str, pathlib.Path, list]] = None,
+        val_folder: Optional[Union[str, pathlib.Path, list]] = None,
+        test_folder: Optional[Union[str, pathlib.Path, list]] = None,
+        predict_folder: Optional[Union[str, pathlib.Path, list]] = None,
+        extensions: Optional[Tuple[str, ...]] = None,
+    ):
+        super().__init__(
+            train_data=train_folder,
+            val_data=val_folder,
+            test_data=test_folder,
+            predict_data=predict_folder,
+        )
+
+        self.extensions = extensions
+
+    @staticmethod
+    def find_classes(dir: str) -> Tuple[List[str], Dict[str, int]]:
+        """
+        Finds the class folders in a dataset. Ensures that no class is a subdirectory of another.
+
+        Args:
+            dir: Root directory path.
+
+        Returns:
+            tuple: (classes, class_to_idx) where classes are relative to (dir), and class_to_idx is a dictionary.
+        """
+        classes = [d.name for d in os.scandir(dir) if d.is_dir()]
+        classes.sort()
+        class_to_idx = {cls_name: i for i, cls_name in enumerate(classes)}
+        return classes, class_to_idx
+
+    def load_data(self, data: Any, dataset: Optional[Any] = None) -> Iterable[Mapping[str, Any]]:
+        classes, class_to_idx = self.find_classes(data)
+        if not classes:
+            files = [os.path.join(data, file) for file in os.listdir(data)]
+            return [{
+                'input': file
+            } for file in filter(
+                lambda file: has_file_allowed_extension(file, self.extensions),
+                files,
+            )]
+        else:
+            self.preprocess.set_state(LabelsState(classes))
+        dataset.num_classes = len(classes)
+        data = make_dataset(data, class_to_idx, extensions=self.extensions)
+        return [{'input': input, 'target': target} for input, target in data]
+
+
+class FilesDataSource(DataSource, ABC):
+
+    def __init__(
+        self,
+        train_files: Optional[Union[Sequence[Union[str, pathlib.Path]], Iterable[Union[str, pathlib.Path]]]] = None,
```

Comment:
same as before, I would simplify to primitive types the inputs

Reply:
Yeah, this should just be sequence of str